### PR TITLE
PERF-3488: Performance tests for variadic aggregate expressions

### PR DESCRIPTION
**Pull request by Adityav369**

### CODE REVIEW

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (mtandrei) [inline src/workloads/execution/VariadicAggregateExpressions.yml:157]: [nit] prefer numbers to be written out (so, "Ten" instead of "10")

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (mtandrei) [inline src/workloads/execution/VariadicAggregateExpressions.yml:238]: I'm not sure that we want to test let? If we do though, we should make sure that the "in" expression is small and doesn't reference most of the bound fields. That way, we can measure the cost of binding variables.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (mtandrei) [inline src/workloads/execution/VariadicAggregateExpressions.yml:268]: we should also have coverage for the set expressions ($setUnion/equals/intersection)

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (mtandrei) [inline src/workloads/execution/VariadicAggregateExpressions.yml:5]:
```suggestion
  can take in a variable number of expressions as arguments.
```

**Review by mtandrei — CHANGES_REQUESTED** [review]

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done. I use 5 and 10 now.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml:61]: As per offline discussion, since we have only 10 fields, I chose not to use the ^cycle generator. Let me know if you think I should use it and reduce the number of fields we insert in a document.

Comment (mtandrei) [reply src/workloads/execution/VariadicAggregateExpressions.yml:61]: There's two questions here:
- How many args should we have? 5 and 10 seem better, but I'm not sure whether something like 3 and 6 or 3 and 5 would be more realistic. I've pinged Chris and Katya on slack to see if they have any opinions; if they don't let's stick with with 5 and 10
- What should the document schema look like? The current schema feels a bit crazy, but it's probably for the best as a "realistic" use case wouldn't duplicate field arguments. However, one (very optional) suggestion: could you try and use the [object generator to cut down on some of the redundancy here?](https://github.com/mongodb/genny/blob/dcd3a0527cce7afbe1aaec90b3cb488672253b5e/src/workloads/docs/Generators.yml#L236) If not, it's no big deal, but it could make this more parseable.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (mtandrei) [inline src/workloads/execution/VariadicAggregateExpressions.yml:147]: discussing offline how we might simplify this further

**Review by mtandrei — CHANGES_REQUESTED** [review]

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml:159]: As per offline discussion, I made changes to the generator by adding a deterministic mode to Choose, and now we just control the parameter "number" when constructing an array.

**Review by rtimmons — CHANGES_REQUESTED** [review]
Making `^Choose` have a `deterministic` option is an interesting concept I never would have thought of. I'm curious how useful it will be in the general sense. Glad to see it as an option.

A few suggestions to decrease chances of mis-configs and increase test-coverage to guard against future regressions. Please push back wherever you disagree. Only real block from my perspective is adding some mutual-exclusion for other options if `deterministic` is specified.

Note I didn't extensively review your workload yamls. I'll leave that to your team and patch-tests. In general I'd prefer a separate PR for the valgen vs the workload yaml so reverts (if needed) can be more surgical, but no biggie this time.

Thanks for pushing on this. Glad you were able to avoid a confusing double-`^Concat` situation.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (Adityav369) [reply src/value_generators/src/DocumentGenerator.cpp:1924]: Done. If deterministic and weights are both specified, it fails with and InvalidSyntaxException. I checked for other configs, but looks like it is just these weights and deterministic.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (Adityav369) [reply src/value_generators/src/DocumentGenerator.cpp]: Ah yes! Done.

File: src/value_generators/test/DocumentGeneratorTestCases.yml
Comment (Adityav369) [reply src/value_generators/test/DocumentGeneratorTestCases.yml:123]: Done.

**Review by Adityav369 — COMMENTED** [review]
Thanks Mihai and Ryan for going over my PR! I have implemented the suggested changes.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (mtandrei) [inline src/value_generators/src/DocumentGenerator.cpp:529]: This looks like extra whitespace; please delete it (I'm not sure whether there's a cpp linter for genny or not)

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (mtandrei) [reply src/workloads/execution/VariadicAggregateExpressions.yml:159]: Nice work on this; resolving!

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (mtandrei) [inline src/workloads/execution/VariadicAggregateExpressions.yml:26]: can we use the new deterministic ^Choose to auto generate these? As in, something like ^Join together 'fieldname' and a choose array from 1 to 10? 

Similar comment for the document schema, but if you're going to do that, it might be worth leaving a comment describing the schema.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (mtandrei) [inline src/workloads/execution/VariadicAggregateExpressions.yml:325]: Similar comment, can we automatically generate these fields instead of hard coding them? One idea is to bind the referenced fields by name, but then auto-generate random fields for the non referenced fields.

**Review by mtandrei — CHANGES_REQUESTED** [review]
Nice work Adi! Almost L-G-T-M, just a couple more suggestions and this should be good to go.

I only skimmed the cpp changes; they look good, but I will defer to Ryan for the final sign offs.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (Adityav369) [reply src/value_generators/src/DocumentGenerator.cpp]: Done.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (mtandrei) [inline src/value_generators/src/DocumentGenerator.cpp:576]: Would it make sense for 'deterministic' to be a boolean value?

File: src/value_generators/test/DocumentGeneratorTestCases.yml
Comment (mtandrei) [inline src/value_generators/test/DocumentGeneratorTestCases.yml:119]:
```suggestion
  - Name: Generate Array with choose deterministic where 'number' is greater than size of 'from'
```

File: src/workloads/docs/Generators.yml
Comment (mtandrei) [inline src/workloads/docs/Generators.yml:232]: [nit] s/from/'from' (generally, prefer to have named arguments in single quotes). Here and elsewhere.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (mtandrei) [inline src/workloads/execution/VariadicAggregateExpressions.yml:27]: I guess it's not possible to use ^Inc and get a string, but: the lists of 50 and 100 integers ought to be simplified further . Could this be done by joining two deterministic choose lists; something like ["", "1", "2", "3", "4"] and the list of 10, then ["", "1", "2", "3", "4",  "5", "6", "7", "8", "9"]  and the list of 10? This would make it so that we start at 0, but this is definitely worth it if we can avoid hard coding a list.

**Review by mtandrei — APPROVED** [review]
LGTM with some final comments. Great work Adi!

**Review by rtimmons — DISMISSED** [review]
Nicely done!

File: src/value_generators/src/DocumentGenerator.cpp
Comment (Adityav369) [reply src/value_generators/src/DocumentGenerator.cpp]: Done.

File: src/value_generators/test/DocumentGeneratorTestCases.yml
Comment (Adityav369) [reply src/value_generators/test/DocumentGeneratorTestCases.yml]: Done.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (mtandrei) [inline src/workloads/execution/VariadicAggregateExpressions.yml:42]: one problem that I ran into with ^Inc is that for using it repeatedly, you will keep getting higher and higher values ([see comment here](https://github.com/mongodb/genny/blob/b41ef95d4870c8d375a0ae3f7a99b44bd02bb643/src/workloads/execution/FilterWithComplexLogicalExpression.yml#L98-L111)). Before merging, please verify that this produces the fields that you expect.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (mtandrei) [inline src/workloads/execution/VariadicAggregateExpressions.yml:42]:
```suggestion
  IntGenerator: &IntGenerator {^Join: {array: [{^FormatString: {format: "%03d", withArgs: [{^Inc: {start: 1}}]}}]}}
```

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (mtandrei) [inline src/workloads/execution/VariadicAggregateExpressions.yml:261]:
```suggestion
# every argument of $ifNull.
```

**Review by mtandrei — APPROVED** [review]
Still LGTM

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml:42]: I was running into that issue initially. I used ^FixedGeneratedValue to generate the list of ints, so it is generated only once.

File: src/workloads/execution/VariadicAggregateExpressions.yml
Comment (Adityav369) [reply src/workloads/execution/VariadicAggregateExpressions.yml]: Done.

**Review by mtandrei — APPROVED** [review]
Still LGTM

**Pull request merged**
